prow/cmd/build/controller.go

```diff
@@ -17,46 +17,424 @@ limitations under the License.
+	return metav1.Now()
+}
+
+var (
+	groupVersionKind = schema.GroupVersionKind{
+		Group:   prowjobv1.SchemeGroupVersion.Group,
+		Version: prowjobv1.SchemeGroupVersion.Version,
+		Kind:    "ProwJob",
+	}
+)
+
+// reconcile ensures that the prowjob and build
+func reconcile(c reconciler, key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource %s: %v", key, err))
+	}
+
+	var wantBuild bool
+
+	pj, err := c.getProwJob(namespace, name)
+	switch {
+	case apierrors.IsNotFound(err):
+		// Do not want build
+	case err != nil:
+		return fmt.Errorf("get prowjob: %v", err)
+	case pj.Spec.Agent != prowjobv1.BuildAgent:
+		// Do not want a build for this job
+	case pj.DeletionTimestamp == nil:
+		wantBuild = true
+	}
+
+	var haveBuild bool
+
+	b, err := c.getBuild(namespace, name)
+	switch {
+	case apierrors.IsNotFound(err):
+		// Do not have a build
+	case err != nil:
+		return fmt.Errorf("get build: %v", err)
+	case b.DeletionTimestamp == nil:
+		haveBuild = true
+	}
+
+	// Should we create or delete this build?
+	switch {
+	case !wantBuild:
+		if !haveBuild {
+			logrus.Infof("Deleted %s", key)
+			return nil
+		}
+		switch or := metav1.GetControllerOf(b); {
+		case or == nil, or.APIVersion != groupVersionKind.GroupVersion().String(), or.Kind != groupVersionKind.Kind:
+			return nil // Not controlled by this
+		}
+		logrus.Infof("Delete builds/%s", key)
+		if err = c.deleteBuild(namespace, name); err != nil {
+			return fmt.Errorf("delete build: %v", err)
+		}
+		return nil
+	case finalState(pj.Status.State):
+		logrus.Infof("Finished %s", key)
+		return nil
+	case wantBuild && pj.Spec.BuildSpec == nil:
+		return errors.New("nil BuildSpec")
+	case wantBuild && !haveBuild:
+		if b, err = makeBuild(*pj); err != nil {
+			return fmt.Errorf("make build: %v", err)
+		}
+		logrus.Infof("Create builds/%s", key)
+		if b, err = c.createBuild(namespace, b); err != nil {
+			return fmt.Errorf("create build: %v", err)
+		}
+	}
+
+	// Ensure spec is correct
+	pj.Spec.BuildSpec.Generation = b.Spec.Generation
+	if !reflect.DeepEqual(b.Spec, *pj.Spec.BuildSpec) {
+		logrus.Infof("\n%#v\n!=\n%#v", b.Spec, *pj.Spec.BuildSpec)
+		b2 := b.DeepCopy()
+		b2.Spec = *pj.Spec.BuildSpec
+		logrus.Infof("Update builds/%s", key)
+		if b, err = c.updateBuild(namespace, b2); err != nil {
+			return fmt.Errorf("update build spec: %v", err)
+		}
+	}
+
+	// Ensure prowjob status is correct
+	haveState := pj.Status.State
+	haveMsg := pj.Status.Description
+	wantState, wantMsg := prowJobStatus(b.Status)
+	if haveState != wantState || haveMsg != wantMsg {
+		npj := pj.DeepCopy()
+		if npj.Status.StartTime.IsZero() {
+			npj.Status.StartTime = c.now()
+		}
+		if npj.Status.CompletionTime.IsZero() && finalState(wantState) {
+			now := c.now()
+			npj.Status.CompletionTime = &now
+		}
+		npj.Status.State = wantState
+		npj.Status.Description = wantMsg
+		logrus.Infof("Update prowjobs/%s", key)
```

Comment:
You should use a finalizer on the Build to ensure that this controller has time to update the ProwJob before a Build is removed. I add a finalizer to my CRD using a mutating admission webhook in `ci-vm-operator`